soc: xtensa: split the cavs server and the FW loader

enjiamai opened the pull request:
The PR is to spilt the cavs server and FW loader, to make the 
cavstool.py a pure FW loader and runner as it used to be. 
The reasons to do this is:

- Keep the FW loader won't be affected by the changes of
   the client-server-based HW service as possible. It also
   make the debugging more easier.
- Spawn another process for the FW loader. If the FW loader
   got stuck during executing, the service can stop it.

Fixes #47652. 
 
And change the server program's name to remote-fw-service.py
because we also try make the client-server-based FW service 
being able to use on other boards besides cavs boards in the future.

kv2019i reviewed (dismissed):
Not on expert on the client-server part, but changes to cavstool.py look good. Thanks!

ceolin commented on soc/xtensa/intel_adsp/tools/hw-service.py:
```diff
@@ -0,0 +1,449 @@
+                            'intel_adsp_cavs20', 'intel_adsp_cavs20_jsl',
+                            'intel_adsp_cavs25', 'intel_adsp_cavs25_tgph']
+
+        board_x86_hypervisor = ['acrn_ehl_crb', 'acrn']
```
isn't it cavs specific ?

enjiamai replied:
En...I made it also support other boards, but it seems not good to put them in the this PR/commit will put them in another PR . Thanks for the review!

ceolin replied:
it is still here ... if this file is supposed to be used for other targest other than intel_adsp, it has to be moved to a better place

enjiamai replied:
Ok, got it. @ceolin , yes, I made this intel_adsp only now.

ceolin commented on soc/xtensa/intel_adsp/tools/hw-service.py:
```diff
@@ -0,0 +1,449 @@
+
+        elif self.board in board_x86_hypervisor:
+            if not load_cmd:
+                self.load_cmd = "./acrn-test-pty.exp"
```
This is not even in tree, is it ?

enjiamai replied:
> I like the separation but instead of copying the code in one commit and deleting in a later commit (which is not even the next commit), can you please MOVE as much code as possible in a single commit? This will help git track the moved lines of code and it will also make the code review and git log easier to follow.

Yes, I will fix this. Thanks!

> Please don't mix the addition of brand new features in the same commit that copies or moves a lot of code https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes Actually, for Continuous Integration reasons please separate these in different Pull Requests.

You are right. I should do it this way. I will keep changes of CAVS only in this PR, and support for other boards should be another PR.

> Instead of adding a lock (and try to track IP addresses?), why not just accept a single TCP connection at a time?
Add a lock is trying to prevent two users to run west/twister from accessing the board at the same time. Because we have the "request" and "log" service, these two services are occupied by a single user in a time period. I not sure if a single TCP connection can solve this. I will try it.

> I don't understand how separating the code into 2 different files can "fix" a ROM boot failure, this does not make sense to me sorry, can you elaborate? So far I've seen two totally different kinds of boot failures:

Sorry for making confusing here. The major difference is I try to spawn another process to run the pure cavstool.py, and if it is stuck, the sever can kill it and continue to run the next test. It's kind of like what the original cavstool.py and cavstwist.sh did.
Another simple fix is to add a timeout in the "waiting for ROM starting" while loop: 
```
    log.info(f"Wait for ROM startup, ADSPCS = 0x{dsp.ADSPCS:x}")
    time.sleep(1)
    while (dsp.SRAM_FW_STATUS >> 24) != 5: pass
```
Otherwise it would stuck here forever and block all following tests.
I was considering to add this earlier, but finally I thought the best way is to totally split the server and the fw_loader. It seems like I have added too much code in this cavstool.py and might confuse the debugging of the tool or fw_loader itself. How do you think?

> * Intermittent firmware or firmware authentication issues. These cannot be "fixed" by moving Python code around. Python code can at best perform retries or other workarounds but it cannot "solve" anything
> * Lack of delimiter and other misuse of raw TCP sockets in discussed in [#46864 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/46864#issuecomment-1167545791) and links. These _can_ be fixed in Python but not by shuffling code around files.
I will continue on improving this. Indeed your suggestion is right.

> Not a fan of the new filename `hw-service.py` but I don't have any great alternative to offer. Maybe `remote-fw-service.py`?
Good suggestion. It sounds better. I will update it.

ceolin replied:
????

enjiamai replied:
Yes, remove it.

ceolin commented on soc/xtensa/intel_adsp/tools/hw-service.py:
```diff
@@ -0,0 +1,449 @@
+                sys.exit(1)
+
+            if not self.load_cmd:
+                self.load_cmd = "/usr/bin/picocom"
```
This dependency should be listed somewhere. I think most of the project uses `minicom`, can't you use it to be more aligned with the rest ?

enjiamai replied:
This is for boards which output log to the serial port. You are right, and I will using the python's serial instead of calling a picocom or minicom. (And it has remove from this PR, the support for other boards will be added in another PR)

ceolin reviewed (dismissed):
I am confuse with this mix of cavs and hypervisors

marc-hb reviewed (dismissed):
I like the separation but instead of copying the code in one commit and deleting in a later commit (which is not even the next commit), can you please MOVE as much code as possible in a single commit? This will help git track the moved lines of code and it will also make the code review and git log easier to follow.


Please don't mix the addition of brand new features in the same commit that copies or moves a lot of code https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes
Actually, for Continuous Integration reasons please separate these in different Pull Requests.


Instead of adding a lock (and try to track IP addresses?), why not just accept a single TCP connection at a time?


I don't understand how separating the code into 2 different files can "fix" a ROM boot failure, this does not make sense to me sorry, can you elaborate? So far I've seen two totally different kinds of boot failures:
- Intermittent firmware or firmware authentication issues. These cannot be "fixed" by moving Python code around. Python code can at best perform retries or other workarounds but it cannot "solve" anything
- Lack of delimiter and other misuse of raw TCP sockets in discussed in https://github.com/zephyrproject-rtos/zephyr/issues/46864#issuecomment-1167545791 and links. These _can_ be fixed in Python but not by shuffling code around files.


Not a fan of the new filename `hw-service.py` but I don't have any great alternative to offer. Maybe `remote-fw-service.py`?

enjiamai commented:
> I like the separation but instead of copying the code in one commit and deleting in a later commit (which is not even the next commit), can you please MOVE as much code as possible in a single commit? This will help git track the moved lines of code and it will also make the code review and git log easier to follow.

Yes, I will fix this. Thanks!

> Please don't mix the addition of brand new features in the same commit that copies or moves a lot of code https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes Actually, for Continuous Integration reasons please separate these in different Pull Requests.

You are right. I should do it this way. I will keep changes of CAVS only in this PR, and support for other boards should be another PR.

> Instead of adding a lock (and try to track IP addresses?), why not just accept a single TCP connection at a time?
Add a lock is trying to prevent two users to run west/twister from accessing the board at the same time. Because we have the "request" and "log" service, these two services are occupied by a single user in a time period. I not sure if a single TCP connection can solve this. I will try it.

> I don't understand how separating the code into 2 different files can "fix" a ROM boot failure, this does not make sense to me sorry, can you elaborate? So far I've seen two totally different kinds of boot failures:

Sorry for making confusing here. The major difference is I try to spawn another process to run the pure cavstool.py, and if it is stuck, the sever can kill it and continue to run the next test. It's kind of like what the original cavstool.py and cavstwist.sh did.
Another simple fix is to add a timeout in the "waiting for ROM starting" while loop: 
```
    log.info(f"Wait for ROM startup, ADSPCS = 0x{dsp.ADSPCS:x}")
    time.sleep(1)
    while (dsp.SRAM_FW_STATUS >> 24) != 5: pass
```
Otherwise it would stuck here forever and block all following tests.
I was considering to add this earlier, but finally I thought the best way is to totally split the server and the fw_loader. It seems like I have added too much code in this cavstool.py and might confuse the debugging of the tool or fw_loader itself. How do you think?

> * Intermittent firmware or firmware authentication issues. These cannot be "fixed" by moving Python code around. Python code can at best perform retries or other workarounds but it cannot "solve" anything
> * Lack of delimiter and other misuse of raw TCP sockets in discussed in [#46864 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/46864#issuecomment-1167545791) and links. These _can_ be fixed in Python but not by shuffling code around files.
I will continue on improving this. Indeed your suggestion is right.

> Not a fan of the new filename `hw-service.py` but I don't have any great alternative to offer. Maybe `remote-fw-service.py`?
Good suggestion. It sounds better. I will update it.

enjiamai commented:
I appreciate you review! @marc-hb , @ceolin , I am working on addressing your comments.

marc-hb commented:
> Add a lock is trying to prevent two users to run west/twister from accessing the board at the same time. Because we have the "request" and "log" service, these two services are occupied by a single user in a time period.

OK, but why would you allow two different users from the same IP address?

> I not sure if a single TCP connection can solve this. I will try it.

Grab a lock as the very first thing in the `handle()` method and this should be enough?

> if it is stuck, the sever can kill it and continue to run the next test. 

OK, this is not fixing any boot issue, this is better error handling for it (which is great).

marc-hb commented:
> Grab a lock as the very first thing in the handle() method and this should be enough?

Make the lock acquisition either wait with a long timeout or fail immediately depending on what kind of user interface you want to offer. Python also has a `request_queue_size` that could be interesting.

teburd reviewed (commented):
Really only looked over cavstool.py, that looks fine now to me and its quite nice to see it not include the socket server

enjiamai committed:
soc: xtensa: tool: duplicate the cavstool.py to remote-fw-service.py

Copy cavstool.py to remote-fw-service.py in order to track changes
for following splitting work of cavstool.py.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai committed:
soc: xtensa: split the cavs server and the FW loader

Spilt the cavs server and FW loader, to make the cavstool.py
a pure FW loader and runner as it used to be. The reasons is
try to keep the FW loader won't be affected by the client-
server-based HW service as possible, and more easy to debug.

After splitting them, the service program spawns another
process to run the FW loader. The advantage is no matter when
the FW loader is stuck, the service can detect and stop it,
then continue to next test.

Fixes #47652.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai committed:
soc: xtensa: remove the serivce logic in the FW loader

Remove the remote service logic in the FW loader, now
the FW loader is a pure FW loader as it used to be.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai committed:
soc: xtensa: cavstool: add support for logs coming from HDA logging

1) HDA logging seems to be using some padding like '\x00'. Such
string can print well on a terminal but corrupt the string match.
And this can cause false failure if RunID matching is affected.
Remove such padding before checking RunID.

An affected RunID example:
'7aa9ba3c6db12\x00\...\x00\x00\x00\x00d0c7fcf382a4af40ec6'

Expected:
'7aa9ba3c6db12d0c7fcf382a4af40ec6'

2) Use non-displayable chars for live connection check. Otherwise
the log output will have subtle garbage like extra spaces which
locate randomly.

This solution comes from the PR #50071, the author is smrtos.

Signed-off-by: Ming Shao <ming.shao@intel.com>
Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai committed:
boards: xtensa: update the documentation for changes of cavstool

Update the usage for cavstool.py to remote-fw-service.py to adapt
the renaming. And also correct the params --log-port and --req-port
in the doc.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

enjiamai commented:
Hi @marc-hb , @ceolin  sorry for late reply, I spent some time to update and tested it over and over, it seems quite stable now. I think it is ready for your review again. 

@marc-hb , your suggestion is right. I remove the the "new feature" (the locked for resource conflict for multiple users and support for other boards such as x86, etc) and try to make this PR only being focus on "split FW loader" and bug fix.
So the lock feature will be put in next PR. 

And I try to make the commits stone alone as possible (maybe it is still a little large for each commits). Also trying to make the change of the cavstool.py traceable, do you think is it ok for you?

I appreciate your review again!

enjiamai commented:
Hi @nashif , @MaureenHelm , @teburd,  @kv2019i , @laurenmurphyx64 ,@smrtos, could you please help with reviewing it? Thank you! 
Besides fix the https://github.com/zephyrproject-rtos/zephyr/issues/47652, also I rarely see the sporadic connection failed when running parallelly this on 3 boards after applying this PR.

enjiamai commented:
> OK, but why would you allow two different users from the same IP address? 

I guess the judge from IP address might be enough in our user scenario, and client(west/twister) log and request services often comes from the same IP. Sorry @marc-hb , I am not very sure did I get your point or not? 

> Grab a lock as the very first thing in the `handle()` method and this should be enough?
> Make the lock acquisition either wait with a long timeout or fail immediately depending on what kind of user interface you want to offer. Python also has a request_queue_size that could be interesting.

Yes, I think it is ok to put it in the handle() and refine this according to your suggestion when the "lock feature" being added in the next PR, I believe the lock feature is important for us because we have many developers sharing the intel_adsp boards.
> 
> OK, this is not fixing any boot issue, this is better error handling for it (which is great).

Thanks for clarifying it, this description is more accuracy.

teburd commented on boards/xtensa/intel_adsp_cavs15/doc/index.rst:
```diff
@@ -85,7 +86,7 @@ loader script as root:
-   cavs15$ sudo ./cavstool.py
+   cavs15$ sudo ./remote-fw-service.py
 
 Cavstool_server.py is a daemon which accepts a firmware image from a remote host
```
remote-fw-service.py

teburd commented on boards/xtensa/intel_adsp_cavs15/doc/index.rst:
```diff
@@ -77,6 +77,7 @@ ip address), and the user account is ``user``. Then copy the python tool to the
+   $ scp boards/xtensa/intel_adsp/tools/remote-fw-service.py user@cavs15:
 
 
 Note that the ``/dev/hda`` device file created by the diagnostic driver must
```
The diagnostic driver isn't needed

teburd reviewed (approved):
Looks ok, could use with some python lint checking and correcting, and a few doc updates

I'm pretty sure a python linter wouldn't be happy with class x_y_z(): type naming for example.

Would recommend a follow up round doing so.
